libraries/testing/tester.cpp

```diff
@@ -529,11 +529,19 @@ namespace eosio::testing {
    signed_block_ptr base_tester::produce_blocks( uint32_t n, bool empty ) {
       signed_block_ptr res;
       if( empty ) {
-         for( uint32_t i = 0; i < n; ++i )
+         for( uint32_t i = 0; i < n; ++i ) {
+            // for performance, only vote on the last four to move finality
+            // This is 4 instead of 3 because the extra block has to be produced to log_irreversible
+            if (n > 4)
+               control->allow_voting(i >= n - 4);
             res = produce_empty_block();
+         }
```

Comment:
I think it would be nicer to have a single loop 
```suggestion
         for( uint32_t i = 0; i < n; ++i ) {
            // for performance, only vote on the last four to move finality
            // This is 4 instead of 3 because the extra block has to be produced to log_irreversible
            if (n > 4)
               control->allow_voting(i >= n - 4);
            res = empty ? produce_empty_block() : produce_block();
         }
```